File: src/main/java/com/gusrylmubarok/reddit/backend/service/impl/VoteServiceImpl.java
package com.gusrylmubarok.reddit.backend.service.impl;

import com.gusrylmubarok.reddit.backend.dto.VoteDto;
import com.gusrylmubarok.reddit.backend.exceptions.BackendRedditException;
import com.gusrylmubarok.reddit.backend.exceptions.PostNotFoundException;
import com.gusrylmubarok.reddit.backend.model.Post;
import com.gusrylmubarok.reddit.backend.model.Vote;
import com.gusrylmubarok.reddit.backend.repository.PostRepository;
import com.gusrylmubarok.reddit.backend.repository.VoteRepository;
import com.gusrylmubarok.reddit.backend.service.AuthService;
import com.gusrylmubarok.reddit.backend.service.VoteService;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

import static com.gusrylmubarok.reddit.backend.model.VoteType.UPVOTE;


@Service
@AllArgsConstructor
public class VoteServiceImpl implements VoteService {
    private final VoteRepository voteRepository;
    private final PostRepository postRepository;
    private final AuthService authService;

    @Override
    @Transactional
    public void vote(VoteDto voteDto) {
        Post post = postRepository.findById(voteDto.getPostId())
                .orElseThrow(() -> new PostNotFoundException("Post Not Found with Id-"
                        + voteDto.getPostId()));
        Optional<Vote> voteByPostAndUser = voteRepository.findTopByPostAndUserOrderByVoteIdDesc(post, authService.getCurrentUser());
        if (voteByPostAndUser.isPresent() &&
                voteByPostAndUser.get().getVoteType().equals(voteDto.getVoteType())) {
            throw new BackendRedditException("You have already "
                    + voteDto.getVoteType() + "'d for this post");
        }
        if (UPVOTE.equals(voteDto.getVoteType())) {
            post.setVoteCount(post.getVoteCount() + 1);
        } else {
            post.setVoteCount(post.getVoteCount() - 1);
        }
        voteRepository.save(mapToVote(voteDto, post));
        postRepository.save(post);
    }

    private Vote mapToVote(VoteDto voteDto, Post post) {
        return Vote.builder()
                .voteType(voteDto.getVoteType())
                .post(post)
                .user(authService.getCurrentUser())
                .build();
    }
}

File: src/main/java/com/reddot/service/impl/VoteServiceImpl.kt
package com.reddot.service.impl

import com.reddot.entity.Post
import com.reddot.entity.Vote
import com.reddot.entity.VoteEnum
import com.reddot.exception.BadRequestException
import com.reddot.model.MessageResponse
import com.reddot.model.VoteDto
import com.reddot.repository.PostRepository
import com.reddot.repository.VoteRepository
import com.reddot.service.AuthService
import com.reddot.service.VoteService
import org.springframework.stereotype.Service
import java.util.*

@Service
class VoteServiceImpl(
    private val voteRepository: VoteRepository,
    private val postRepository: PostRepository,
    private val authService: AuthService
) : VoteService {
    override fun vote(voteDto: VoteDto): MessageResponse {
        val post = postRepository.findById(voteDto.postId)
        if (post.isEmpty) {
            throw BadRequestException("not found with post ${voteDto.postId}")
        }

        val voteByPostAndUser = voteRepository.findTopByPostAndUserOrderByIdDesc(post.get(), authService.getCurrentUser())
        if (voteByPostAndUser.isPresent && voteByPostAndUser.get().voteType == voteDto.voteType) {
            throw BadRequestException("you have already ${voteDto.voteType}' for this post")
        }

        if (VoteEnum.UPVOTE == voteDto.voteType) {
            post.get().voteCount = post.get().voteCount + 1
        } else {
            post.get().voteCount = post.get().voteCount - 1
        }
        voteRepository.save(mapToVote(voteDto, post.get()))
        postRepository.save(post.get())

        return MessageResponse("You ${voteDto.voteType}")
    }

    private fun mapToVote(voteDTO: VoteDto, post: Post): Vote {
        return Vote(
            id = null,
            voteType = voteDTO.voteType,
            post = post,
            user = authService.getCurrentUser(),
            createdAt = Date(),
            updatedAt = Date()
        )
    }
}